More robust task messaging

mvdbeek opened the pull request:
This implements a database heartbeat and allows the QueueWorker to receive and send tasks to local and remote processes as long as they are connected to the database. This should immediately improve things like the data table reload that is triggered after running a data manager tool and therefore fixes https://github.com/galaxyproject/galaxy/issues/7564 and https://github.com/galaxyproject/galaxy/issues/7568

I'm also implementing a way to wait and get the results of a task, which I will use for waiting for these reloads (in a new PR). Longer term we could use this to manage tool dependencies for handlers that run remotely and scheduling the config watching to a single process (instead of every process doing that separately). 
Everything implemented here has unit tests for amqp connections via sqlite and postgresql as well as a real message broker.

mvdbeek committed:
Add simple test for queue worker

mvdbeek committed:
Fix unit tests, need to retry sending message

mvdbeek committed:
Also test with postgresql fixture

mvdbeek commented:
I remember now what the problem is, no fanout / broadcast for the sqlachemy (aka in-memory) transport, so queues / exchanges need to be declared up front, which makes this not very dynamic when spinning up new handlers. For uwsgi that's not too bad because we call `UWSGIApplicationStack.worker()`, which should give us all the uwsgi workers, but classic webless handlers need to be registered manually or can't be reached via control tasks.

Now I guess workers could also report themselves with an alive message in the database, which would close that gap. I'll see if that could be done.

mvdbeek committed:
Make processes and mules report heartbeat in database

It's important to not set the DatabaseHeartbeat `server_name` prior to forking,
which we avoid be getting `server_name` from
`application_stack.app.config.server_name`.
`WorkerProcess.server_name` for uwsgi web processes will look like:
`main.web.1` and mules like `main.job-handlers.1` while webless workers
remain named `handler1`.

mvdbeek committed:
Add option to return worker results

Occasionally we want to wait for a toolbox or data table
reload before continuing, so knowing when that's done
(=wait for the result) is helpful.

Getting the result of a task seems like a good idea,
we could use this for a whole lot of things.
One thing is targeting functions to (remote) handlers
that have access to information that web handlers don't,
like how busy a cluster is or which dependencies are
available or whether a handler is dead.

mvdbeek committed:
Drop unreachable code

mvdbeek committed:
Use database heartbeat info for building queues on-the-fly

mvdbeek committed:
Add test using multiple independent apps

Unfortunately adding a third app doesn't seem to work.
In that case only the last 2 apps receive the control task.

I do know that this isn't a problem with real separate processes,
but I don't know what is causing this.

mvdbeek committed:
Delete remaining messages on startup

So far I don't think we have any messages that should survive
restarts. If we do at some point we should route them differently
(or change the existing worker to use ephemeral messages).

mvdbeek committed:
Avoid KeyError on rabbitmq transports

We apparently are no guaranteed a `reply-to` Key.

mvdbeek committed:
Also test against rabbitmq

mvdbeek committed:
Make tests use finalizer instead of yield

and verify tasks get executed the expected number of times.
Since the part after the yield statement runs to early it
was shutting down the app(s) prematurely.

mvdbeek committed:
Clean up unused options in QueueWorker and ControlTask

mvdbeek committed:
Only run rabbitmq queue worker test if GALAXY_TEST_AMQP_INTERNAL_CONNECTION is set

mvdbeek commented:
We did actually use postgres for the integration tests, it seems ...

nsoranzo commented on lib/galaxy/dependencies/pipfiles/default/pinned-dev-requirements.txt:
```diff
@@ -34,6 +34,7 @@ pyparsing==2.3.1
 pytest-html==1.20.0
 pytest-metadata==1.8.0
 pytest-pythonpath==0.7.3
+pytest-postgresql==1.3.4
```
You need to add this to `lib/galaxy/dependencies/pipfiles/default/Pipfile`

mvdbeek replied:
Thanks @nsoranzo, I've rebased that into https://github.com/galaxyproject/galaxy/pull/7548/commits/e7e8283560ef061450a96c96d7cb2f61a6d41ba4

mvdbeek committed:
Use ApplicationStack for accessing server name

mvdbeek committed:
Fully initialize empty databaseses if GALAXY_TEST_DBURI is set and DB doesn't exist

mvdbeek committed:
Add tables that were only added as part of migrations

in e19092e9c52d740d88ac53622438c7fc7edb72fb

mvdbeek commented:
It's all green and ready to review again!

nsoranzo commented on lib/galaxy/model/migrate/versions/0151_add_worker_process.py:
```diff
@@ -0,0 +1,53 @@
+    _drop(WorkerProcess_table)
+
+
+def _create(table):
```
FYI, I'm refactoring these methods in https://github.com/galaxyproject/galaxy/pull/7621 to avoid the code duplications.

dannon commented:
@mvdbeek Ran into a snag testing the administrative job locking under uwsgi, digging deeper to see if it's a wonky configuration on my end or something else.

mvdbeek commented:
Just tried this too, it locks the jobs but the UI doesn't show the lock, so you can't unlock, right ?
fwiw I tested this with the tool whitelist, which was working fine. I can look at that and also use the new methods from #7621

mvdbeek committed:
Need to build the control queues after forking

Fixes https://github.com/galaxyproject/galaxy/pull/7548#issuecomment-478125862

mvdbeek committed:
Reduce migration boilerplate

mvdbeek commented:
I think I fixed it with https://github.com/galaxyproject/galaxy/pull/7548/commits/4493d126d6fe1c99c32ab77f87b873e598680f25 -- moving that up in the init from `bind_and_start` was one of the last things I did ...

dannon commented:
@mvdbeek Cool, thanks, I'll check it out again.

mvdbeek committed:
Allow a little more time for thread to start up in heartbeat unit test

mvdbeek committed:
Re-add fix

https://jenkins.galaxyproject.org/job/docker-integration-py3/145/testReport/junit/test.integration.test_mules_as_job_handlers/MultipleMulesAsJobHandlersIntegrationTestCase/test_runs_on_mule/
looks like
http://dev.list.galaxyproject.org/UnmappedClassError-Class-kombu-transport-sqlalchemy-Message-is-not-mapped-td4667071.html
which was fixed in 81f864626a96cb50b3b943e7441a00dd9668dbcd

mvdbeek commented:
@galaxybot test py3

dannon reviewed (approved):
@mvdbeek Thanks for this; the changes look great and it's working well in my testing so far -- let's get it into more hands! :)

natefoo commented:
@mvdbeek I think this does not address the issue in #5601 but correct me if I'm wrong?

> I discovered this in the course of figuring out how some of the uWSGI workers on my front end VM 1 were receiving `rebuild_toolbox_search_index` async tasks via the message queue before they had themselves reloaded the toolbox and queued the task; it's because the workers on VM 2 had already done it, 2's workers had queued the tasks, and because they _both_ have workers with `server_name`s of `main.web.{0..7}`, 1 was picking up those tasks that 2 had intended for itself. So it should be documented somewhere that `server_name` will need to include a hostname-unique element if you use the same internal message queue with multiple hosts. Alternatively, we can automatically include some kind of host identifier when those exchanges are set up.

(I'm not saying this PR should have fixed that issue, just wanted to make sure I should not close it).

Since servers now register themselves via heartbeat and my server names overlap there may be new issues? But I will finally fix things so I don't have overlapping server names.

mvdbeek commented:
Oh yeah, that's right. I could also add a uuid-postfix or hostname or really anything else to the task queue if that would make things easier ?

natefoo commented:
Hmm. Something host-specific (`hostid(1)`?) would work for my case, I'm trying to think of any other cases where we might need something unique to some other property and can't think of any.

Isn't this specific to the silly way we send messages meant for all processes though? In reality we should use a fanout exchange instead of generating server names, I'd think. So it'd probably make more sense if I just make my base server names unique until we implement that?

mvdbeek commented:
It is, but fanout exchanges don't work with in-memory transports like sqlalchemy. I think docker-based handlers will also benefit from including the hostname by default.

natefoo commented:
I can't think of a downside, so if you're volunteering to spend a few minutes to add it, be my guest, but it's certainly not a dire need. :smile:

mvdbeek commented:
I don't wanna bring main down (again) so I'll do it 😆

natefoo commented:
Thanks!

blankenberg commented:
I had an instance that I had to comment out `pytest-postgresql==1.3.4` to get to start up. The machine it is on doesn't have `pg_config` , and pytest-postgresql was causing a new version of `psycopg2-binary==2.8.1` be attempted to be built (which requires `pg_config`). 

The venv already has a working version of `psycopg2-binary==2.7.4` that is being used (and is required for Galaxy as an optional dependency).

mvdbeek commented:
We can update to 1.4 which removed the hard requirement on psycopg2, but why is your instance pulling in dev-requirements ? Are you setting DEV_WHEELS ?

blankenberg commented:
Not sure exactly, it is started with
`DRMAA_LIBRARY_PATH=/usr/local/lib/libdrmaa.so sh run.sh --daemon`

mvdbeek commented:
k, I think https://github.com/galaxyproject/galaxy/pull/7703 should fix that additional dependency, if you wanna give it a try.

nsoranzo commented:
@blankenberg Then dev wheels should get installed only if you are on the `dev` branch.

blankenberg commented:
Yes, definitely on a dev-based branch there (realtimetools).

blankenberg commented:
Will try the change shortly, once I can restart the server again.
